sysctl: Add more sysctl action

rhn-jaeshin opened the pull request:
kernel.hung_task_warnings, vm.max_map_count, vm.zone_reclaim_mode

Signed-off-by: Jay Shin <jaeshin@redhat.com>

ryran commented on xsos:
```diff
@@ -2855,6 +2855,7 @@ SYSCTL() {
   __P kernel.random.entropy_avail "[bits] "
   __P kernel.hung_task_panic "[bool] "
   __Pa kernel.hung_task_timeout_secs ""  '{if ($1>0) printf "\"%s\"%s  (secs task must be D-state to trigger)", $1, H0; else printf "\"0\"%s  (khungtaskd disabled)", H0}'
+  __Pa kernel.hung_task_warnings "[0-10] "  '{if ($1==0) printf "\"%s\"%s  (hung has previously occurred for more than 10 times)", $1, H0; else if ($1<10) printf "\%s\%s  (hung has previously occurred more than once)", $1, H0; else if ($1>10) printf "\%s\%s  (Default hung_task_warnings value has changed)", $1, H0; else printf "\%s\%s  (hung has never occurred)", $1, H0}'
```
Hey @rhn-jaeshin thanks for the contribution. Sorry for neglecting it for so long. Breaking this line down...

Looking at kernel-doc I see:

> hung_task_warnings:
> 
> The maximum number of warnings to report. During a check interval
> if a hung task is detected, this value is decreased by 1.
> When this value reaches 0, no more warnings will be reported.
> This file shows up if CONFIG_DETECT_HUNG_TASK is enabled.
> 
> -1: report an infinite number of warnings.

While the above doesn't say it, from my checking, it appears to default to `10`.

So it's writable but it also automatically changes based on things happening on the system. What a weird sysctl. My interpretation from that:

* you can set it to `-1` and then it will stay at that static value and all hung-tasks will be reported
* you can set it to `0` and it will stay at that static value and no hung-tasks will be reported
* you can set it to a positive value (or leave at default of `10`), but then it will be automatically decremented by one each time a hung-task occurs ... and they'll be reported each time until it gets to zero, after which point they won't be reported any more

How bizarre. Unless you know the original setting, there's really no way to glean good information from the current value (unless it's `-1`). BLEH. So yeah not sure how to capture that in our guidance, but I'll give it a shot.

rhn-jaeshin commented:
Thank Ryan for your opinions.
I hope you can find out the best option to indicate the hung information.


2019년 4월 4일 (목) 오후 1:08, Ryan Sawhill Aroha <notifications@github.com>님이 작성:

> *@ryran* commented on this pull request.
> ------------------------------
>
> In xsos <https://github.com/ryran/xsos/pull/240#discussion_r272003431>:
>
> > @@ -2855,6 +2855,7 @@ SYSCTL() {
>    __P kernel.random.entropy_avail "[bits] "
>    __P kernel.hung_task_panic "[bool] "
>    __Pa kernel.hung_task_timeout_secs ""  '{if ($1>0) printf "\"%s\"%s  (secs task must be D-state to trigger)", $1, H0; else printf "\"0\"%s  (khungtaskd disabled)", H0}'
> +  __Pa kernel.hung_task_warnings "[0-10] "  '{if ($1==0) printf "\"%s\"%s  (hung has previously occurred for more than 10 times)", $1, H0; else if ($1<10) printf "\%s\%s  (hung has previously occurred more than once)", $1, H0; else if ($1>10) printf "\%s\%s  (Default hung_task_warnings value has changed)", $1, H0; else printf "\%s\%s  (hung has never occurred)", $1, H0}'
>
> Hey @rhn-jaeshin <https://github.com/rhn-jaeshin> thanks for the
> contribution. Sorry for neglecting it for so long. Breaking this line
> down...
>
> Looking at kernel-doc I see:
>
> hung_task_warnings:
>
> The maximum number of warnings to report. During a check interval
> if a hung task is detected, this value is decreased by 1.
> When this value reaches 0, no more warnings will be reported.
> This file shows up if CONFIG_DETECT_HUNG_TASK is enabled.
>
> -1: report an infinite number of warnings.
>
> While the above doesn't say it, from my checking, it appears to default to
> 10.
>
> So it's writable but it also automatically changes based on things
> happening on the system. What a weird sysctl. My interpretation from that:
>
>    - you can set it to -1 and then it will stay at that static value and
>    all hung-tasks will be reported
>    - you can set it to 0 and it will stay at that static value and no
>    hung-tasks will be reported
>    - you can set it to a positive value (or leave at default of 10), but
>    then it will be automatically decremented by one each time a hung-task
>    occurs ... and they'll be reported each time until it gets to zero, after
>    which point they won't be reported any more
>
> How bizarre. Unless you know the original setting, there's really no way
> to glean good information from the current value (unless it's -1). BLEH.
> So yeah not sure how to capture that in our guidance, but I'll give it a
> shot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ryran/xsos/pull/240#pullrequestreview-222540107>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AqLE-VSRYJy7cWgYwtfnS6_LwlNNC3Boks5vdWyigaJpZM4aEZMl>
> .
>

ryran commented on xsos:
```diff
@@ -2919,15 +2920,17 @@ SYSCTL() {
   __P vm.swappiness "[0-100] "
-  __P vm.min_free_kbytes ""
   __P vm.vfs_cache_pressure "[0-100] "
+  __P vm.zone_reclaim_mode "[0-4] "
```
@rhn-jaeshin From my reading of the kernel-docs, it's not accurate to say `[0-4]`. Looks like it's a bitmask from `0` to `7`.

ryran replied:
EDIT: Ahhhhh digging further, it seems it really is a bitmask. So you can set it to 0, 1, 3, 5, or 7.

ryran commented on xsos:
```diff
@@ -2919,15 +2920,17 @@ SYSCTL() {
   __P vm.dirty_expire_centisecs
   __P vm.dirty_writeback_centisecs
+  __P vm.max_map_count ""
+  __P vm.min_free_kbytes ""
```
@rhn-jaeshin Heh... I don't particularly care, but what's your rationale for moving `min_free_kbytes` from below to up here?

rhn-jaeshin commented:
I moved vm.min_free_kbytes with alphabetical order. It does not matter
where it is.

Anyway, regarding hung_task_warnings, it can be fixed by the way.

-  __Pa kernel.hung_task_warnings "[0-10] "  '{if ($1==0) printf "\"%s\"%s
(hung has previously occurred for more than 10 times)", $1, H0; else if
($1<10) printf "\%s\%s  (hung has previously occurred more than once)", $1,
H0; else if ($1>10) printf "\%s\%s  (Default hung_task_warnings value has
changed)", $1, H0; else printf "\%s\%s  (hung has never occurred)", $1, H0}'
+  __Pa kernel.hung_task_warnings "[0-10] "  '{if ($1<0) printf "\%s\%s
(report an infinite number of warnings)", $1, H0; else if ($1==0) printf
"\"%s\"%s  (hung has previously occurred for more than 10 times)", $1, H0;
else if ($1<10) printf "\%s\%s  (hung has previously occurred more than
once)", $1, H0; else if ($1>10) printf "\%s\%s  (Default hung_task_warnings
value has changed)", $1, H0; else printf "\%s\%s  (hung has never
occurred)", $1, H0}'

zone_reclaim_mode : yes you are wright, it contains 8 combination from 0 to
7
https://access.redhat.com/solutions/2209921



2019년 4월 4일 (목) 오후 1:19, Ryan Sawhill Aroha <notifications@github.com>님이 작성:

> *@ryran* commented on this pull request.
> ------------------------------
>
> In xsos <https://github.com/ryran/xsos/pull/240#discussion_r272004875>:
>
> > @@ -2919,15 +2920,17 @@ SYSCTL() {
>    __Pa vm.dirty_background_bytes ""  '{if ($1>0) printf "\"%s\"%s  (%.1f MiB)", $1, H0, $1/1024/1024; else printf "\"0\"%s  (disabled -- check dirty_background_ratio)", H0}'
>    __P vm.dirty_expire_centisecs
>    __P vm.dirty_writeback_centisecs
> +  __P vm.max_map_count ""
> +  __P vm.min_free_kbytes ""
>
> @rhn-jaeshin <https://github.com/rhn-jaeshin> Heh... I don't particularly
> care, but what's your rationale for moving min_free_kbytes from below to
> up here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ryran/xsos/pull/240#pullrequestreview-222541828>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AqLE-QqwbeqKiZYKO_ig4QNzyWaWYxwQks5vdW88gaJpZM4aEZMl>
> .
>

rhn-jaeshin commented:
You can discuss this value with Super-Jamie.


On Thu, 4 Apr 2019 at 13:54, Ryan Sawhill Aroha <notifications@github.com>
wrote:

> *@ryran* commented on this pull request.
> ------------------------------
>
> In xsos <https://github.com/ryran/xsos/pull/240#discussion_r272009270>:
>
> >    __P vm.vfs_cache_pressure "[0-100] "
> +  __P vm.zone_reclaim_mode "[0-4] "
>
> EDIT: Ahhhhh digging further, it seems it really is a bitmask. So you can
> set it to 0, 1, 3, 5, or 7.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ryran/xsos/pull/240#discussion_r272009270>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AqLE-SrlOoKzXmi2GN8Vwfuww50UFHIpks5vdXddgaJpZM4aEZMl>
> .
>

ryran commented:
Dealt with in #241 

Thanks @rhn-jaeshin
